tools/bin/mbedtls-rewrite-branch-style

```diff
@@ -0,0 +1,586 @@
+        if path.suffix in {'.c', '.h'}:
+            return True
+        if dir_chain == ('tests', 'suites') and path.suffix == '.function':
+            return True
+        if dir_chain == ('scripts', 'data_files') and path.suffix == '.fmt':
+            return True
+        return False
+
+    def is_file_exempt(self, path: pathlib.PurePath) -> Optional[str]:
+        """Whether path is exempt from restyling.
+
+        Return the reason why path is exempt (a non-empty string),
+        or None if path is not exempt.
+        """
+        if path.parts[0] == '3rdparty':
+            return '3rdparty'
+        # Hard-coded list of files that are generated in 2.28.
+        # In development, generated files are not checked into Git.
+        if path.parts in self.GENERATED_FILES_2_28:
+            return 'generated'
+        return None
+
+    def filter_styled_files(self, filenames: List[str]) -> List[str]:
+        """Return filenames with non-restylable files removed."""
+        keep = []
+        for filename in filenames:
+            path = pathlib.PurePath(filename)
+            if not self.is_c_file(path):
+                continue
+            why_exempt = self.is_file_exempt(path)
+            if why_exempt:
+                self.info('Exempt from restyling ({}): {}', why_exempt, filename)
+                continue
+            self.info('Will restyle: {}', filename)
+            keep.append(filename)
+        return keep
+
+    def restyle_files(self, filenames: List[str]) -> None:
+        """Restyle the given files in place."""
+        if not filenames:
+            return
+        for num in range(2):
+            self.info('Pass #{} of uncrustify', num)
+            subprocess.check_call([self.UNCRUSTIFY_EXE,
+                                   '-c', self.CONFIG_FILE,
+                                   '--no-backup'] +
+                                  filenames)
+
+    def restyle_commit_onto_current(self, new_commit: str) -> None:
+        """Apply a restyled content of new_commit onto HEAD."""
+        self.info('Applying {} onto {}', new_commit, self.head_sha())
+        # We want the new commit's metadata (author, commit message),
+        # the new commit's content (but restyled), and the old commit's
+        # history. This is a cherry-pick operation, with some custom
+        # way to merge the contents. We could perhaps define a custom merge
+        # strategy, but that seems hard (especially since we do want to
+        # tweak the content even if there are no merge conflicts). So we
+        # pick some method of resolving the conflict automatically
+        # that ensures that added and removed files are according to the
+        # cherry-picked commit, but might mess up modified files. Then
+        # we override the content of modified files, and finally we restyle
+        # files that need restyling.
+        self.run_git(['cherry-pick', '-Xtheirs', '--allow-empty', new_commit])
```

Comment:
This line also has the potential to run into an empty cherry-pick, especially if is run in a branch that already has the new style. Which that is clearly a user error case, it may be worth to put the cherry-pick in a try: except block.

Reply:
I prefer to end up with an empty commit than to try and catch empty-commit errors without accidentally catching other errors.

Ah, I see why `--allow-empty` isn't enough:

> `--allow-empty` (…) Note also, that use of this option only keeps commits that were initially empty (i.e. the commit recorded the same tree as its parent). Commits which are made empty due to a previous commit are dropped. To force the inclusion of those commits use --keep-redundant-commits.

So I'm adding `--keep-redundant-commits` and hoping that this won't hide problems that I'm not thinking of.

Reply:
Done in https://github.com/Mbed-TLS/mbedtls-docs/pull/73